src/app/options.cpp

```diff
@@ -0,0 +1,394 @@
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ *
+ * Contact : chris@qbittorrent.org
+ */
+
+#include "options.h"
+
+#include <iostream>
+
+#include <QDebug>
+#include <QFileInfo>
+#include <QProcessEnvironment>
+#include <QTextStream>
+
+#ifdef Q_OS_WIN
+#include <QMessageBox>
+#endif
+
+#include "base/utils/misc.h"
+
+namespace
+{
+    // Base option class. Encapsulates name operations.
+    class Option
+    {
+    public:
+        QString usage() const
+        {
+            QString res;
+            if (m_shortcut)
+                res += QLatin1String("-") + QLatin1Char(m_shortcut) + QLatin1String(" | ");
+            res += fullParameter();
+            return padUsageText(res);
+        }
+
+    protected:
+        constexpr Option(const char *name, char shortcut = 0)
+            : m_name {name}
+            , m_shortcut {shortcut}
+        {
+        }
+
+        QString fullParameter() const
+        {
+            return QLatin1String("--") + QLatin1String(m_name);
+        }
+
+        QString shortcutParameter() const
+        {
+            return QLatin1String("-") + QLatin1Char(m_shortcut);
+        }
+
+        bool hasShortcut() const
+        {
+            return m_shortcut != 0;
+        }
+
+        QString envVarName() const
+        {
+            return QLatin1String("QBT_") +
+                   QString(QLatin1String(m_name)).toUpper().replace(QLatin1Char('-'), QLatin1Char('_'));
+        }
+
+        static QString padUsageText(const QString &usage)
+        {
+            const int TAB_WIDTH = 8;
+            QString res = QLatin1String("\t") + usage;
+            if (usage.size() < 2 * TAB_WIDTH)
+                return res + QLatin1String("\t\t");
+            else
+                return res + QLatin1String("\t");
+        }
+
+    private:
+        const char *m_name;
+        const char m_shortcut;
+    };
+
+    // Boolean option.
+    class BoolOption: public Option
+    {
+    public:
+        constexpr BoolOption(const char *name, char shortcut = 0)
+            : Option {name, shortcut}
+        {
+        }
+
+        bool operator == (const QString &arg) const
+        {
+            return (hasShortcut() && ((arg.size() == 2) && (arg == shortcutParameter())))
+                   || (arg == fullParameter());
+        }
+
+        bool value(const QProcessEnvironment &env) const
+        {
+            QString val = env.value(envVarName());
+            // we accept "1" and "true" (upper or lower cased) as boolean 'true' values
+            return (val == QLatin1String("1") || val.toUpper() == QLatin1String("TRUE"));
+        }
+    };
+
+    inline bool operator == (const QString &s, const BoolOption &o)
+    {
+        return o == s;
+    }
+
+    // Option with string value. May not have a shortcut
+    struct StringOption: protected Option
```

Comment:
`BoolOption` is an exception (with  its public inheritance). But probably I just need to move `Option::usage()` from `Option` to `BoolOption`.